### PR TITLE
[ODS-4252] Upgrade Load Tools Project to netcore 3.1

### PR DESCRIPTION
**Pull request by BhanuprakashThunga**

### CODE REVIEW

File: Utilities/DataLoading/EdFi.LoadTools.Test/EdFi.LoadTools.Test.csproj [hunk 1]
```diff
@@ -42,22 +42,24 @@
...
+      <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.6.1" />
+      <PackageReference Include="Moq" Version="4.14.3" />
+      <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+      <PackageReference Include="NUnit" Version="3.12.0" />
+      <PackageReference Include="NUnit3TestAdapter" Version="3.17.0">
           <PrivateAssets>all</PrivateAssets>
           <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
         </PackageReference>
-        <PackageReference Include="RestSharp" Version="105.2.3" />
+      <PackageReference Include="RestSharp" Version="106.11.4" />
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/EdFi.LoadTools.Test.csproj:63]: do we need restsharp with the introduction of fubarcoder.restsharpt.portable.core?

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs [hunk 2]
```diff
@@ -55,7 +61,7 @@ public class ApiGetTests
               }
             }}";
 
-        private static readonly string Address = ConfigurationManager.AppSettings["TestingWebServerAddress"];
+        private static readonly string Address = "http://localhost:23456/";
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs:64]: this should come from the config

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs [hunk 3]
```diff
@@ -75,78 +81,99 @@ public class ApiGetTests
         private readonly IOAuthTokenHandler tokenHandler = Mock.Of<IOAuthTokenHandler>();
 
         private Resource _resource;
-        private IDisposable _webApp;
 
         [OneTimeSetUp]
         public void Setup()
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs:86]: should we be using aync task here?

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs [hunk 4]
```diff
@@ -75,78 +81,99 @@ public class ApiGetTests
...
+                                        else
+                                        {
+                                            await context.Response.WriteAsync(JsonConvert.SerializeObject(_data));
+                                        }
+                                    }));
+                        })
+                ;
+
+            Host = hostBuilder.Build();
+            Host.StartAsync();
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs:158]: should this be await Host.StartAsync()?

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs [hunk 5]
```diff
@@ -75,78 +81,99 @@ public class ApiGetTests
...
         }
 
         [OneTimeTearDown]
         public void Cleanup()
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs:162]: should this be async Task? Cleanup()

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs [hunk 6]
```diff
@@ -75,78 +81,99 @@ public class ApiGetTests
...
         {
-            _webApp.Dispose();
+            Host.StopAsync();
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs:164]: should this be await Host.StopAsync()?

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs [hunk 7]
```diff
@@ -75,78 +81,99 @@ public class ApiGetTests
...
         public void GetAll_should_store_results_in_dictionary()
         {
             var dictionary = new Dictionary<string, JArray>();
 
-            var subject = new GetAllTest(_resource, dictionary, _configuration, tokenHandler);
+            var client = Host.GetTestClient();
+            client.BaseAddress = new System.Uri(Address);
+
+            var subject = new GetAllTest(_resource, dictionary, _configuration, tokenHandler, client);
             var result = subject.PerformTest().Result;
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs:177]: should this be await subject.PerformTest() ?

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs [hunk 8]
```diff
@@ -75,78 +81,99 @@ public class ApiGetTests
...
         [OneTimeTearDown]
         public void Cleanup()
         {
-            _webApp.Dispose();
+            Host.StopAsync();
+            Host.Dispose();
         }
 
         [Test]
         public void GetAll_should_store_results_in_dictionary()
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs:169]: these tests should be async Task as the type

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs [hunk 9]
```diff
@@ -156,12 +183,11 @@ public void GetAll_should_store_results_in_dictionary()
...
-                             {
-                                 [ResourceName] = _data
-                             };
+            var dictionary = new Dictionary<string, JArray> {[ResourceName] = _data};
+            var client = Host.GetTestClient();
+            client.BaseAddress = new System.Uri(Address);
 
-            var subject = new GetAllSkipLimitTest(_resource, dictionary, _configuration, tokenHandler);
+            var subject = new GetAllSkipLimitTest(_resource, dictionary, _configuration, tokenHandler, client);
             var result = subject.PerformTest().Result;
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs:191]: this should be await subject.PerformTest()

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs [hunk 10]
```diff
@@ -6,37 +6,53 @@
...
 {
     [TestFixture]
     public class GetDependenciesTests
     {
         [Test]
-        public void Should_succeed_against_a_running_server()
+        public async System.Threading.Tasks.Task Should_succeed_against_a_running_serverAsync()
         {
-            var address = ConfigurationManager.AppSettings["TestingWebServerAddress"];
+            var address = "http://localhost:23456/";
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs:26]: this needs to come from the config

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs [hunk 11]
```diff
@@ -6,37 +6,53 @@
...
+            client.BaseAddress = new System.Uri(address);
+
+            var configuration = Mock.Of<IApiMetadataConfiguration>(cfg => cfg.Url == address);
+            var subject = new GetStaticDependenciesTest(configuration, client);
+            var result = subject.PerformTest().Result;
+            Assert.IsTrue(result);
         }
 
         [Test]
         public void Should_fail_against_no_server()
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs:51]: this test should be of type async Task

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs [hunk 12]
```diff
@@ -6,37 +6,53 @@
...
+                                    await ctx.Response.WriteAsync("successful result")));
+                    });
+
+            var host = await hostBuilder.StartAsync();
+            var client = host.GetTestClient();
+            client.BaseAddress = new System.Uri(address);
+
+            var configuration = Mock.Of<IApiMetadataConfiguration>(cfg => cfg.Url == address);
+            var subject = new GetStaticDependenciesTest(configuration, client);
+            var result = subject.PerformTest().Result;
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs:46]: this should be an await subject.PerformTest()

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs [hunk 13]
```diff
@@ -6,37 +6,56 @@
...
     public class GetVersionTests
     {
         [Test]
-        public void Should_succeed_against_a_running_server()
+        public async System.Threading.Tasks.Task Should_succeed_against_a_running_serverAsync()
         {
-            var address = ConfigurationManager.AppSettings["TestingWebServerAddress"];
+            //TODO should be replaced with json
+            // var address = ConfigurationManager.AppSettings["TestingWebServerAddress"];
+            var address = "http://localhost:23456/";
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs:28]: again this should come from the config

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs [hunk 14]
```diff
@@ -6,37 +6,56 @@
...
+                                    await ctx.Response.WriteAsync("successful result")));
+                    });
+
+            var host = await hostBuilder.StartAsync();
+            var client = host.GetTestClient();
+            client.BaseAddress = new System.Uri(address);
+
+            var configuration = Mock.Of<IApiMetadataConfiguration>(cfg => cfg.Url == address);
+            var subject = new GetStaticVersionTest(configuration, client);
+            var result = subject.PerformTest().Result;
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs:49]: this should be await subject.PerformTest()

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs [hunk 15]
```diff
@@ -6,37 +6,56 @@
...
+            Assert.IsTrue(result);
         }
 
         [Test]
         public void Should_fail_against_no_server()
         {
             const string Url = "http://localhost:12345";
             var configuration = Mock.Of<IApiMetadataConfiguration>(cfg => cfg.Url == Url);
-            var subject = new GetStaticVersionTest(configuration);
+            var subject = new GetStaticVersionTest(configuration, null);
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs:58]: that is the change here? why the null?

File: Utilities/DataLoading/EdFi.BulkLoadClient.Console/EdFi.BulkLoadClient.Console.csproj [hunk 16]
```diff
@@ -37,10 +37,10 @@
   </ItemGroup>
   <ItemGroup>
     <PackageReference Include="EdFi.Suite3.Common" Version="5.0.0" />
-    <PackageReference Include="FluentCommandLineParser" Version="1.4.3" />
-    <PackageReference Include="log4net" Version="2.0.8" />
+    <PackageReference Include="FluentCommandLineParser-netstandard" Version="1.4.3.13" />
+    <PackageReference Include="log4net" Version="2.0.10" />
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
-    <PackageReference Include="RestSharp" Version="105.2.3" />
+    <PackageReference Include="RestSharp" Version="106.11.4" />
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.BulkLoadClient.Console/EdFi.BulkLoadClient.Console.csproj:43]: do we need restharp here?

**Review by jamessmckay — CHANGES_REQUESTED** [review]

**Comment by jamessmckay** [comment]
In general the tests where we are using .Result as a suffix should be converted to async Task as the type from void.

File: Utilities/DataLoading/EdFi.LoadTools.Test/EdFi.LoadTools.Test.csproj (hunk 1 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/EdFi.LoadTools.Test.csproj]: Removed RestSharp

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs (hunk 2 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs]: Replaced with AppSettings.json

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs (hunk 3 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs]: Changed to Async task

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs (hunk 4 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs]: Updated

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs (hunk 5 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs]: Updated to async task

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs (hunk 6 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs]: Updated

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs (hunk 7 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs]: Updated

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs (hunk 8 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs]: Updated

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs (hunk 9 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/APIGetTests.cs]: Updated

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs (hunk 10 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs]: Now reading form appsettings.json

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs (hunk 11 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs]: Modified to async task

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs (hunk 12 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetDependenciesTests.cs]: updated to use await

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs (hunk 13 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs]: Added appsettings.json

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs (hunk 14 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs]: Updated

File: Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs (hunk 15 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.LoadTools.Test/SmokeTests/GetVersionTests.cs]: Added default null parameter to the constructor

File: Utilities/DataLoading/EdFi.BulkLoadClient.Console/EdFi.BulkLoadClient.Console.csproj (hunk 16 above)
Comment (BhanuprakashThunga) [reply Utilities/DataLoading/EdFi.BulkLoadClient.Console/EdFi.BulkLoadClient.Console.csproj]: Removed restsharp

**Review by jamessmckay — CHANGES_REQUESTED** [review]
Broken build on this pull request
```
4
Step 9/11: Run Smoke Tests (PowerShell)
10:30:44
  PowerShell Executable: C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
10:30:44
  Working directory: D:\BuildAgent\work\904029b3fc87a31e\Ed-Fi-ODS-Implementation
10:30:44
  Command: C:\Windows\sysnative\WindowsPowerShell\v1.0\powershell.exe
10:30:44
  PowerShell arguments: -NoProfile, -NonInteractive, -ExecutionPolicy, ByPass, -File, D:\BuildAgent\temp\buildTmp\powershell8355071349416250200.ps1
10:30:45
  Using repositories: Ed-Fi-ODS, Ed-Fi-ODS-Implementation
10:30:45
  10:30:45
  Path                                                         
10:30:45
  ----                                                         
10:30:45
  D:\buildagent\work\904029b3fc87a31e\Ed-Fi-ODS-Implementation\
10:30:45
  10:30:45
  10:30:45
  10:30:45
  apiAppConfig              D:\buildagent\work\904029b3fc87a31e\Ed-Fi-ODS-Implementation\Application\EdFi.Ods.Api.Integra
10:30:45
                            tionTestHarness\bin\**\appsettings.json                                                      
10:30:45
  apiClientNameSandbox      ODS/API                                                                                      
10:30:45
  apiKey                    smoke                                                                                        
10:30:45
  apiNamespaceUri           http://edfi.org                                                                              
10:30:45
  apiSecret                 smokeSecret                                                                                  
10:30:45
  apiUrlBase                http://localhost:8765                                                                        
10:30:45
  apiUrlData                http://localhost:8765/data/v3                                                                
10:30:45
  apiUrlMetadata            http://localhost:8765/metadata                                                               
10:30:45
  apiUrlOAuth               http://localhost:8765                                                                        
10:30:45
  apiYear                                                                                                                
10:30:45
  buildConfiguration        release                                                                                      
10:30:45
  bulkLoadTempJsonConfig    D:\BuildAgent\temp\buildTmp\smokeTestconfig.json                                             
10:30:45
  loadToolsSolution         D:\buildagent\work\904029b3fc87a31e\Ed-Fi-ODS\Utilities\DataLoading\LoadTools.sln            
10:30:45
  msbuild_exe               G:\Program Files (x86)\Microsoft Visual                                                      
10:30:45
                            Studio\2019\BuildTools\MSBuild\Current\bin\amd64\MSBuild.exe                                 
10:30:45
  noExtensions              False                                                                                        
10:30:45
  smokeTestDll              D:\buildagent\work\904029b3fc87a31e\Ed-Fi-ODS\Utilities\DataLoading\EdFi.LoadTools.Test\bin\*
10:30:45
                            *\EdFi.OdsApi.Sdk.dll                                                                        
10:30:45
  smokeTestExecutable       D:\buildagent\work\904029b3fc87a31e\Ed-Fi-ODS\Utilities\DataLoading\EdFi.SmokeTest.Console\bi
10:30:45
                            n\**\EdFi.SmokeTest.Console.exe                                                              
10:30:45
  testHarnessAppConfig      D:\buildagent\work\904029b3fc87a31e\Ed-Fi-ODS-Implementation\Application\EdFi.Ods.Api.Integra
10:30:45
                            tionTestHarness\bin\**\appsettings.json                                                      
10:30:45
  testHarnessJsonConfig     D:\buildagent\work\904029b3fc87a31e\Ed-Fi-ODS-Implementation\logistics\scripts\smokeTestHarne
```

File: Utilities/DataLoading/EdFi.SmokeTest.Console/Program.cs [hunk 17]
```diff
@@ -79,13 +87,30 @@ private static void StartCommandLine(Container container)
 
         private static void ConfigureCommandLineContainer(Container container, Configuration configuration)
         {
+            var hostBuilder = new HostBuilder()
```
Comment (jamessmckay) [inline Utilities/DataLoading/EdFi.SmokeTest.Console/Program.cs:90]: confused why we are building a new host here.

File: Utilities/DataLoading/EdFi.LoadTools/ApiClient/SwaggerRetriever.cs [hunk 18]
```diff
@@ -122,11 +127,24 @@ public SwaggerRetriever(IApiMetadataConfiguration configuration)
...
+
+                        webHost.Configure(
+                            app => app.Run(
+                                async ctx =>
+                                    await ctx.Response.StartAsync().ConfigureAwait(false)));
+                    });
+
+            var host = await hostBuilder.StartAsync();
+           
+            using (var client = host.GetTestClient())
```
[the remark this answers is not in the document]
Comment (raghuramakrishna) [reply Utilities/DataLoading/EdFi.LoadTools/ApiClient/SwaggerRetriever.cs]: rolled back code

File: Utilities/DataLoading/EdFi.LoadTools/SmokeTest/CommonTests/GetStaticBaseTest.cs [hunk 19]
```diff
@@ -15,10 +15,12 @@ namespace EdFi.LoadTools.SmokeTest.CommonTests
     public abstract class GetStaticBaseTest : ITest, ITestGenerator
     {
         private readonly string _url;
+        private static HttpClient _httpClient;
 
-        protected GetStaticBaseTest(string url)
+        protected GetStaticBaseTest(string url, HttpClient client = null)
         {
             _url = url;
+            _httpClient = client;
```
[the remark this answers is not in the document]
Comment (raghuramakrishna) [reply Utilities/DataLoading/EdFi.LoadTools/SmokeTest/CommonTests/GetStaticBaseTest.cs]: rolled back code

File: Utilities/DataLoading/EdFi.SmokeTest.Console/Program.cs (hunk 17 above)
Comment (raghuramakrishna) [reply Utilities/DataLoading/EdFi.SmokeTest.Console/Program.cs]: rolled back code

File: Utilities/DataLoading/EdFi.SmokeTest.Console/Program.cs [hunk 20]
```diff
@@ -79,13 +87,30 @@ private static void StartCommandLine(Container container)
...
+            var client = host.GetTestClient();
+
             container.RegisterInstance<IApiConfiguration>(configuration);
             container.RegisterInstance<IApiMetadataConfiguration>(configuration);
             container.RegisterInstance<IOAuthTokenConfiguration>(configuration);
             container.RegisterInstance<IOAuthSessionToken>(configuration);
             container.RegisterInstance<ISdkLibraryConfiguration>(configuration);
             container.RegisterInstance<IDestructiveTestConfiguration>(configuration);
             container.Register<ISmokeTestApplication, SmokeTestApplicationNoBlocks>();
+            container.Register<HttpClient>(() => client, Lifestyle.Singleton);
```
[the remark this answers is not in the document]
Comment (raghuramakrishna) [reply Utilities/DataLoading/EdFi.SmokeTest.Console/Program.cs]: rolled back code

**Comment by jamessmckay** [comment]
Still getting errors when I run the application, in this case I am running the BulkLoadClient using the statemet ```Initialize-MinimalTemplate -samplePath ..\Ed-Fi-Standard\ -noExtensions -noValidation```

**Review by jamessmckay — APPROVED** [review]

**Pull request merged**